### PR TITLE
Adam Dunlap assignment submission

### PR DESCRIPTION
**Pull request by AdamCDunlap**

### CODE REVIEW

File: no-computer.md [hunk 1]
```diff
@@ -58,12 +97,27 @@ making language design decisions), as opposed to "systems" aspects of the
...
+everyone already knows some pretty good designs for them.
+
 
 ### Benefits and drawbacks
 _Why might this be a good idea for a project? Why might this not be a good idea 
 project?_
 
+This would be a good idea because locks are extremely useful and the design of
+them is not too complex. It's not such a good idea because I'm better at
+programming than metalworking.
```
Comment (matthewvalentine) [inline no-computer.md:123]: xD

File: no-computer.md [hunk 2]
```diff
@@ -11,34 +11,65 @@ _What need is met by your idea? Who are you helping? What is that person's
...
+program so that only the ones they trust know it, and the language of a lock
+keeps everyone else out.
 
 ### Why you?
 _What excites you about this idea? How did you come up with it?_
 
+Locks are cool because each one is like its own language. Combinations and keys
+are programs, and these programs have similar structure, but each lock accepts
+a different program. There can also be some languages that accept multiple
+programs, such as a "standard" and "master" key.
```
Comment (matthewvalentine) [inline no-computer.md:36]: How interesting. You've just connecting designing lock types and designing languages. Cool

File: no-computer.md [hunk 3]
```diff
@@ -11,34 +11,65 @@ _What need is met by your idea? Who are you helping? What is that person's
...
+
 
 ### Why a language?
 _Why is a DSL appropriate for your user(s)? How does it address the need?_
 
+Users want to be able to enter a short program that only they know (a
+combination) that will cause the language to grant access, and deny access to
+everyone else. People can easily restrict access to the knowledge of this
+program so that only the ones they trust know it, and the language of a lock
+keeps everyone else out.
```
Comment (matthewvalentine) [inline no-computer.md:28]: Oooo. I would never have thought of this.

File: free-2.md [hunk 4]
```diff
@@ -58,12 +105,25 @@ making language design decisions), as opposed to "systems" aspects of the
...
+could definitely see myself using it.
 
 ### Benefits and drawbacks
 _Why might this be a good idea for a project? Why might this not be a good idea 
 project?_
 
+This is a good idea for a project because it is very language-y and useful.
+There are some drawbacks in that supporting enough different methods of input
+and output could be technically difficult, but that is all part of the language
+design.
```
Comment (matthewvalentine) [inline free-2.md:129]: ATM the current description doesn't sound very language-y to me at all. But, lots of _design_.

You also wouldn't want to get hung up on interface beauty, because that alone is an infinite task.

File: free-2.md [hunk 5]
```diff
@@ -58,12 +105,25 @@ making language design decisions), as opposed to "systems" aspects of the
 project (e.g., implementing a complicated semantics that doesn't require a lot
 of language design)?_
 
+A lot of the time would be spent on figuring out good ways to interact with the
+user. Keeping track of events is pretty easy and notifications can be anywhere
+from easy to pretty difficult. However, the user interaction can also be pretty
+incremental: adding support for manual input would be one level, gmail input
+would be harder, and other methods of input and output could get trickier.
```
Comment (matthewvalentine) [inline free-2.md:112]: For  MVP, you probably wouldn't even need notifications. Just require the user to ask "Hello there, task tracker I've been ignoring because I don't want to face reality. What have I forgotten?"

File: free-2.md [hunk 6]
```diff
@@ -46,6 +89,10 @@ aren't and conjecture why not. If so, describe them and provide links. How well
 do they address the need? Are there any particularly admirable qualities of the
 language? Are there parts of the language you think could be improved?_
 
+There are lots of other DSLs in this domain. Google Inbox, for example, tries
+to combine email and todos, but doesn't have great support for mandatory
+events. I got the idea of importing from email from Google as well, but their
+system doesn't work as well as I hoped.
```
Comment (matthewvalentine) [inline free-2.md:95]: A lot of task-trackers and ticketing systems too. JIRA, Azendoo, Phabricator. And then Todoist, and things like it. Google Calendar is also highly related, but obviously more limited.

File: free-2.md [hunk 7]
```diff
@@ -35,10 +61,27 @@ user? What kinds of errors might occur, and how might they be communicated to
...
 ### Expressiveness
 _What should be easy to do in this language? What should be possible, but
 difficult? What should be impossible or very difficult?_
 
+It should be easy to add both events with defined starting and ending points
+and tasks that have to get done and have an approximate length but no specified
+time frame. Having some other type of event might be possible but difficult,
+although I don't know of anything else necessary. Nothing in the domain should
+really be impossible since the language should be generic enough to support
+basically any tasks.
```
Comment (matthewvalentine) [inline free-2.md:84]: You would probably also want capabilities on par with things like Gantt charts, if feasible. Todoist users have been screaming for that for a long time. Either that, or, allowing such things to be implementable user-side.

File: free-2.md [hunk 8]
```diff
@@ -35,10 +61,27 @@ user? What kinds of errors might occur, and how might they be communicated to
...
+Ideally, there will be some levels of reminders that go from "this is what you
+should take care of today" to "you have to start this assignment now or else
+you'll skip the talk you really want to go to later."
+
+I feel this is the right way because it's how my brain works. When I see a new
+thing I have to do, I only think of it for so long before it leaves my mind so
+I want to get it down as soon as possible. I also am often asking "Do I have
+enough free time to play a few games?" or "I'm feeling productive, what do I
+need to do now?" This system could solve both.
+
```
Comment (matthewvalentine) [inline free-2.md:74]: I agree. This would be great. It would also be nice to be able to associate resources with tasks. For example, "here's a link to the PDF of the reading."

File: free-2.md [hunk 9]
```diff
@@ -11,23 +11,49 @@ _What need is met by your idea? Who are you helping? What is that person's
...
 like? Why is this the right way to interact with the problem domain?_ 
 
+There will have to be several different ways to interact with the language for
+it to be feasible. There has to be easy ways to add events from emails. When an
+email comes in that says there's an interesting-sounding talk on a certain day,
+it should be very easy to put this in, and change settings like how much you
+want to go to it. There should be some similar interface with TODOs, even for
+things such as "deal with this email." Setting up events outside of email
+should also be easy, and it's important to have a nice interface for recurring
+events.
```
Comment (matthewvalentine) [inline free-2.md:56]: How to deal with emails sounds very second-level to me. That is, sort of like the standard library is to Python; the MVP would not include such a thing, but would be directly extendable to it, or capable already of doing it but just not in so few words.

I guess the description of the interface is mostly "similar to TODOs".

What kind of DSL functionality are you thinking of? I think it would be cool if the DSL allowed you to implement your own kinds of functionality. Such as, suppose you didn't support recurring events. But, it was simple enough for a user to make a recipe for such a thing. It could be hard to keep the functionality from getting too powerful though.

File: free-2.md [hunk 10]
```diff
@@ -11,23 +11,49 @@ _What need is met by your idea? Who are you helping? What is that person's
...
 them?_
 
+The need met is people who want to figure out what they need to do at any
+moment and what they can do if they are feeling productive. Currently, there is
+some combination of calendar where events can be manually put, potentially todo
+lists, some of which have different types of deadlines which are all mixed
+together. If this language works out, people can have all of their TODO-like
+things all in one place: things such as mandatory meetings presented one way,
+optional events in another, and hard a soft deadlines that come with some
+weighting so the user knows how important they are.
```
Comment (matthewvalentine) [inline free-2.md:21]: Seems pretty system-heavy, but I have often wanted such a thing.

File: free-1.md [hunk 11]
```diff
@@ -57,12 +98,30 @@ making language design decisions), as opposed to "systems" aspects of the
...
+I would guess that quite a bit of time would go into figuring out exactly how
+the user should interact with the language. This language would be some
+combination of formal logic, general-purpose programming language, and English.
+It needs to be easy to use but also extremely general --- it would be bad if a
+game _almost_ fits in the language because then the language becomes useless.
+Programming the interpreter for the language would also be difficult but
+probably not as bad. If the language is closer to formal logic, some basic
+translation could happen and then be fed into a logic prover. Outputting game
+rules would be complex if one wanted them to look good but simple to get
+something basic done.
```
Comment (matthewvalentine) [inline free-1.md:110]: I think a possible solution here is allowing some amount of "nebulousness". If there's a game state that is hard to describe, just take it as a primitive. The tradeoff is that the checker will not be able to know anything about how one might get into that state, 

However, if the checker IS sophisticated enough to be making complicated checks on things like "this rule is fine because this other state could never happen given ..." then it's just as powerful as a theorem prover.

But, if the difference is just what people express as black boxes or not, I think the general case of your idea would _also_ be as strong as a theorem prover, and moreover as a game _implementation_.  Which means it might be better for it to work more like proof assistants do.

Actually, I think proof assistants in general are a very similar thing as this project is and would be worth looking into.

Incidentally, if you actually want to use formal logic, but perhaps are looking for a reformulation that is more similar to how people actually express their ideas, you could check out Discourse Representation Theory. Probably want to avoid those things as much as possible though.

File: free-1.md [hunk 12]
```diff
@@ -45,6 +78,14 @@ aren't and conjecture why not. If so, describe them and provide links. How well
...
 language? Are there parts of the language you think could be improved?_
 
+I was somewhat inspired by [Ari's CS111 project from
+2014](https://github.com/AriHC/RuleBook), but our ideas are different in that
+his scope was more limited to tic-tac-toe-like games and automatically made a
+playable computer game, while mine will accept more general games and only
+output rules. I also found some other cool game generators like
+[PuzzleScript](http://www.puzzlescript.net/editor.html). But what makes my
+project different is that it can specify arbitrary games, and trades this
+flexibility for the lack of actually implementing them.
```
Comment (matthewvalentine) [inline free-1.md:88]: This may be a hard tradeoff to make.  I think it might even be possible to prove some form of equivalence.  See comment on L110.

File: free-1.md [hunk 13]
```diff
@@ -10,34 +10,67 @@ _What need is met by your idea? Who are you helping? What is that person's
...
 
 ### Expressiveness
 _What should be easy to do in this language? What should be possible, but
 difficult? What should be impossible or very difficult?_
 
+It should be easy to define most games. Simple turn-based card and board games
+will obviously be easier than a complex real-time strategy online game, but it
+should be possible to express both. I can't think of anything in the domain
+that should be really impossible --- ideally, the system is flexible enough to
+define most rule sets.
```
Comment (matthewvalentine) [inline free-1.md:73]: I'm not sure it's quite so straightforward to define most games. Actually, no, I think I've misunderstood your idea a bit.  I guess you are not trying to check also that the logic of the game makes any sense.  That is, you're aiming for the equivalent of a type-checker for game rules. This still seems like it might end up being a pretty hard problem in the general case, but I guess we'll see.

File: free-1.md [hunk 14]
```diff
@@ -10,34 +10,67 @@ _What need is met by your idea? Who are you helping? What is that person's
...
 them?_
 
+My idea helps people who are writing game rules to ensure that they are
+well-written without loopholes. Currently, game makers have to think hard about
+edge cases and not just try to figure out what they are, but how they should be
+handled so that they are most consistent with the rest of the rules. If they
+used a DSL stemming from my idea, they could be notified about what cases are
+left undecided, and the language could automatically help output their rules in
+clear English.
+
```
Comment (matthewvalentine) [inline free-1.md:20]: Ooh. That is brilliant. Immediately I wonder whether such a thing could be only slightly extended to then also give you a game manager program. Although then, if it would not be much additional effort, I wonder if perhaps rigorously defining the logic of a game might end up being a significant challenge. In other words, it may have to be a general-purpose language, because of how variable games might be.

I'm immediately thinking of solutions to these problems. I guess that means this is an excellent idea.

File: anti.md [hunk 15]
```diff
@@ -58,12 +93,24 @@ making language design decisions), as opposed to "systems" aspects of the
...
+the language is the smallest part of the user experience. A ton of time would
+be necessary to parse websites and figure out how to get all of their different
+types of data and updates together.
 
 ### Scope
 _How big an idea is this? How ambitious is this project?_
 
+This is a scalable project. Since the language design aspect is so easy and
+almost all of the time would be devoted to adding website support, the project
+could be toned down or up by just supporting more or fewer websites.
```
Comment (matthewvalentine) [inline anti.md:106]: Depending on how often sites change their formatting, even just maintenance could be a big problem, so there's that.  Except maybe for those sites which provide APIs.

File: anti.md [hunk 16]
```diff
@@ -58,12 +93,24 @@ making language design decisions), as opposed to "systems" aspects of the
 project (e.g., implementing a complicated semantics that doesn't require a lot
 of language design)?_
 
+Very little of their time would be spent on language design because writing in
+the language is the smallest part of the user experience. A ton of time would
+be necessary to parse websites and figure out how to get all of their different
+types of data and updates together.
```
Comment (matthewvalentine) [inline anti.md:99]: True, true.  A big reason why a DSL isn't a good solution / doesn't really solve the problem.

File: anti.md [hunk 17]
```diff
@@ -46,6 +77,10 @@ aren't and conjecture why not. If so, describe them and provide links. How well
 do they address the need? Are there any particularly admirable qualities of the
 language? Are there parts of the language you think could be improved?_
 
+There are some DSLs in this domain, like [RSS](http://www.whatisrss.com/). This
+appear to be somewhat widely supported, but I don't know anyone who uses it. It
+can do well with some things like blogs, but for other types of content it is
+not useful, and the website creator has to specifically support and publish it.
```
Comment (matthewvalentine) [inline anti.md:83]: There are some services like Pushbullet, which also just provide APIs that other people have to use, but that are in common usage.

File: anti.md [hunk 18]
```diff
@@ -11,34 +11,65 @@ _What need is met by your idea? Who are you helping? What is that person's
...
+site changes its format so it can no longer be parsed.
 
 ### Expressiveness
 _What should be easy to do in this language? What should be possible, but
 difficult? What should be impossible or very difficult?_
 
+It should be easy to add most common sites, and possible but difficult to add
+sites that don't have official support, since the user would then have to write
+the parser for that site. It should be impossible to subscribe to sites where
+subscribing doesn't make sense, like sites that never change their content.
```
Comment (matthewvalentine) [inline anti.md:72]: Ah, the way you write parsers might make sense as a DSL. The format you put things into might also.

File: anti.md [hunk 19]
```diff
@@ -11,34 +11,65 @@ _What need is met by your idea? Who are you helping? What is that person's
...
 ### Interface (syntax)
 _How might the user interact with the language? What does programming look 
 like? Why is this the right way to interact with the problem domain?_ 
 
+A user could interact with this language by just writing down what sites they
+are interested in. Possibly there could just be a button that appears on
+supported sites that a user could click to say "I'm interested in this site!"
+This seems like the right way to interact with the domain because actually
+saying what sites you are interested in is a very small part of your time so it
+doesn't matter too much.
```
Comment (matthewvalentine) [inline anti.md:53]: Is something as simple as listing sites really a DSL? On the other hand, perhaps that's a reason why a DSL isn't a good solution: the language part of the problem is very straightforward, and doesn't need the power of a DSL. The language part is "too small" a problem for DSLs.

File: anti.md [hunk 20]
```diff
@@ -11,34 +11,65 @@ _What need is met by your idea? Who are you helping? What is that person's
...
 
 ### Why a language?
 _Why is a DSL appropriate for your user(s)? How does it address the need?_
 
+A DSL could be appropriate because it would specify what the user cares about.
+Unfortunately, I don't think it would work because websites have so many
+different formats that trying to incorporate all of them would be a nightmare.
+A standard user would probably want text posts, images, and videos all in the
+same place, and to do this a large part of the Internet would need to be
+scraped and parsed to get them what they want.
```
Comment (matthewvalentine) [inline anti.md:29]: I don't think this is the reason that a DSL isn't a way to solve this problem, but rather an argument that the problem itself isn't solvable.

File: free-1.md [hunk 21]
```diff
@@ -57,12 +98,30 @@ making language design decisions), as opposed to "systems" aspects of the
...
 ### Benefits and drawbacks
 _Why might this be a good idea for a project? Why might this not be a good idea 
 project?_
 
+This is a good idea for a project because it's clearly a place where a DSL is
+good and could be used for a wide variety of games. Its disadvantages are that
+it's probably not the most useful --- any text that it outputs would probably
+be awkward enough that a game designer would reword it, and that could easily
+reintroduce the very "bugs" that the system is designed to remove.
+
```
Comment (bwiedermann) [inline free-1.md:127]: "Games" is a good domain, but a large one. It wasn't clear at first, but I think you're proposing to make it possible to specify the rules of games and to check those rules for...consistency? Loopholes? Ambiguity? It's not quite clear what would happen when a program runs. Specific examples would help make things clearer, going forward.

Still, I think this is an interesting project idea.  If you go with it, I recommend a bit of research to tie particular kinds of games to, e.g., particular kinds of computable logic. For example, do partial-information games (such as poker) correspond to a particular kind of logic of notion of computability? Also, it might be good at first to limit the domain to a particular style of game, e.g., cards, dice, or boards, then extend the language to handle other styles of games.

You might want to check out [game semantics](https://www.dpmms.cam.ac.uk/~martin/Research/Oldpapers/gamesemantics97scan.pdf). It's more about game theory than games _per se_, but you might find it interesting. There's also a book called [Games, Puzzles, and Computation](http://erikdemaine.org/papers/GPC/).

As a particularly specific suggestion: Could you encode the rules of the logic-based game [Hanabi](https://en.wikipedia.org/wiki/Hanabi_%28card_game%29), such that a program would suggest the best course of action for each turn (including which hint provides the most information)?

File: free-2.md (hunk 4 above)
Comment (bwiedermann) [inline free-2.md:129]: A language (e.g., a composable data structure + algorithms over that structure) may be an appropriate _encoding_ of this domain. I'm not sure that text-based language is the _best_ interface for this domain. I could be wrong (and I'd very much like to be proved so), but I think a text-based interface would be quickly overwhelmed by the complexity of the domain.

The interface(s) you propose may indeed be right on. Alas, they're not linguistic interfaces. If you went ahead with this project, I'd want to make sure that most of your time was on language design, rather than parsing emails, etc. Alternatively, if you wanted to design a syntax that was sort of like natural language--perhaps a stylized agenda that could be sent in an email--that could be interesting and challenging.

It's not yet clear what happens when a program runs, specifically. If you choose this project, it would be good to do some early research on algorithms to compute what you propose.
